src/luxonis_ml/ops/artifacts.py

```diff
@@ -0,0 +1,267 @@
+                if local_checksum != path_to_checksum[path]:
+                    self.dataset._modify_file(path if path != metadata_path else "metadata.parquet")
+            else:
+                self.dataset._add_file(path if path != metadata_path else "metadata.parquet")
+
+        for path in path_to_checksum:
+            if path == metadata_path: path = "metadata.parquet"
+            if path == cache_path: path = ".cache/cache.pickle"
+            if not os.path.exists(f"{self.path}/{path}"):
+                self.dataset._remove_file(path)
+
+
+    def push(self, message=None):
+        if hasattr(self.dataset, 'branch'):
+            t0 = time.time()
+
+            if message is None: message = "luxonis_ml push: default commit message"
+
+            random_str = ''.join(random.choices(string.ascii_letters+string.digits+string.punctuation, k=100))
+            ldf_commit_number = hashlib.sha1(bytes(random_str, 'utf-8')).hexdigest()
+
+            added_and_modified = self.dataset.added_files[self.dataset.bough.value] + self.dataset.modified_files[self.dataset.bough.value]
+
+            if len(added_and_modified) > 0:
+                print("Preparing local data")
+                added_and_modified.append('.cache/cache.pickle') # always push cache
+
+                path_to_info = {}
+                for path in tqdm(added_and_modified):
+                    if path.startswith(self.dataset.bough.value):
+                        split_by = f"{self.dataset.bough.value}/"
+                    else:
+                        split_by = f"/{self.dataset.bough.value}/"
+                    lakefs_path = path.split(split_by)[-1]
+                    relative_path = f"{self.dataset.bough.value}/{lakefs_path}"
+                    checksum = self._get_file_checksum(path)
+                    size = os.path.getsize(path)
+                    s3_path = f"s3://{self.dataset.bucket}/lakefs/{self.name}/_ldf_commits/{ldf_commit_number}/{path.split(split_by)[-1]}"
+                    path_to_info[lakefs_path] = {
+                        's3_path': s3_path,
+                        'checksum': checksum,
+                        'size': size
+                    }
+
+                    if path in ['.cache/cache.pickle', 'metadata.parquet']:
+                        relative_path = path
+
+                    local_dir = relative_path.split(relative_path.split('/')[-1])[0]
+                    if not local_dir.startswith(self.dataset.bough.value):
+                        local_dir = f"{self.dataset.bough.value}/{local_dir}"
+                    os.makedirs(f"tmp/{local_dir}", exist_ok=True)
+                    cmd = f"cp {relative_path} tmp/{local_dir}"
+                    subprocess.check_output(cmd, shell=True)
+
+                print("Uploading data to S3 - this could take some time...")
+                cmd = f"b2 sync tmp/{self.dataset.bough.value} b2://{self.dataset.bucket}/lakefs/{self.name}/_ldf_commits/{ldf_commit_number}/"
```

Comment:
Let's keep it like that for now. Might be worth testing with some other more general CLI that can allow us to provide custom endpoints. In case we switch providers.

Reply:
Would be good to mention this dependency on b2 somewhere in the README.

Reply:
We can also try using rclone for this. I think I found that b2 was faster, but I could explore this more. For now, I'll keep it this way and mention it in the README.

Reply:
Sounds good. Do you remember what the comparison between b2 and rclone was? @conorsim

Reply:
On uploading COCO val:
`b2`: about 11 minutes
`rclone`: about 24 minutes